# 로컬런타임 연결

In [2]:
pip install torch torchvision torchaudio --index-url https://download.pytorch.org/whl/cu118

Looking in indexes: https://download.pytorch.org/whl/cu118
Note: you may need to restart the kernel to use updated packages.


In [2]:
"""로컬 런타임 연결(터미널)
activate tensorflow

jupyter notebook \
    --NotebookApp.allow_origin='https://colab.research.google.com' \
    --port=8888 \
    --NotebookApp.port_retries=0
"""

"로컬 런타임 연결(터미널)\nactivate tensorflow\n\njupyter notebook     --NotebookApp.allow_origin='https://colab.research.google.com'     --port=8888     --NotebookApp.port_retries=0\n"

In [ ]:
# import os
# os.environ['CUDA_LAUNCH_BLOCKING'] = "1"
# os.environ["CUDA_VISIBLE_DEVICES"] = "0"

In [1]:
>>> import sys
>>> print(sys.version)

3.10.14 | packaged by Anaconda, Inc. | (main, May  6 2024, 19:44:50) [MSC v.1916 64 bit (AMD64)]


In [2]:
import pandas as pd
import numpy as np
from tqdm import tqdm
import csv
import matplotlib.pyplot as plt
import torch

print(torch.__version__) #토치 버전 '2.2.2+cu118'
print(torch.version.cuda) #CUDA 버전
print(torch.cuda.get_device_name())
print(torch.cuda.device_count())
print(torch.cuda.is_available())

2.3.0+cu118
11.8
NVIDIA GeForce MX450
1
True


In [3]:
import tensorflow as tf

# Print the installed version of TensorFlow
print("TensorFlow version:", tf.__version__)

# List all available devices detected by TensorFlow
print("Available devices:")
devices = tf.config.list_physical_devices()
for device in devices:
    print(device)

# Check if TensorFlow is currently using a GPU
gpu_available = tf.config.list_physical_devices('GPU')
if gpu_available:
    print("GPU is available for TensorFlow.")
else:
    print("GPU is not available, TensorFlow is using CPU.")

TensorFlow version: 2.10.0
Available devices:
PhysicalDevice(name='/physical_device:CPU:0', device_type='CPU')
PhysicalDevice(name='/physical_device:GPU:0', device_type='GPU')
GPU is available for TensorFlow.


#Mission1-2.

In [4]:
import torch
import torch.nn as nn
import torch.optim as optim
from torch.utils.data import DataLoader
from torchvision import datasets, models, transforms
import os
from tqdm import tqdm  # Progress bar
from torch.utils.data import DataLoader, random_split

C:\Users\SW\.conda\envs\tensorflow\lib\site-packages\tqdm\auto.py:21: TqdmWarning: IProgress not found. Please update jupyter and ipywidgets. See https://ipywidgets.readthedocs.io/en/stable/user_install.html
  from .autonotebook import tqdm as notebook_tqdm


## 이미지 파일 클래스별로 분류하는 코드(재실행할 필요 없음)

In [10]:
import os
import shutil
from tqdm import tqdm

# 원본 이미지 경로와 분류할 경로 설정
source_dir = 'C:/Users/SW/Desktop/데크캠/2024 데이터 크리에이터 캠프 대학부/dataset/training_image'
destination_dir = 'C:/Users/SW/Desktop/데크캠/2024 데이터 크리에이터 캠프 대학부/dataset/sorted_training_images'

# 폴더가 없다면 새로 생성
if not os.path.exists(destination_dir):
    os.makedirs(destination_dir)

# 성별 폴더(W, M)를 생성
gender_dirs = {'W': os.path.join(destination_dir, 'W'), 'M': os.path.join(destination_dir, 'M')}
for gender_dir in gender_dirs.values():
    if not os.path.exists(gender_dir):
        os.makedirs(gender_dir)

# os.listdir로 이미지를 리스트로 가져옴
file_list = os.listdir(source_dir)

# tqdm을 사용하여 진행 상태를 표시
for filename in tqdm(file_list, desc="Sorting images"):
    if filename.endswith('.jpg'):
        # 파일명에서 성별과 클래스를 추출
        parts = filename.split('_')
        gender = parts[-1][0]  # 마지막 부분이 'W' 또는 'M'
        class_name = parts[3]  # 네 번째 부분이 클래스명

        # 성별에 따른 폴더 경로
        gender_dir = gender_dirs.get(gender)

        # 클래스별 폴더 경로 설정
        class_dir = os.path.join(gender_dir, class_name)

        # 클래스별 폴더가 없으면 생성
        if not os.path.exists(class_dir):
            os.makedirs(class_dir)

        # 이미지 파일을 클래스별 폴더로 복사 (파일이 없을 때만 복사)
        source_path = os.path.join(source_dir, filename)
        destination_path = os.path.join(class_dir, filename)
        if not os.path.exists(destination_path):  # 파일이 존재하지 않으면 복사
            shutil.copy(source_path, destination_path)

print("Images have been successfully sorted into gender and class folders.")


Sorting images: 100%|██████████████████████████████████████████████████████████████| 4070/4070 [00:53<00:00, 75.52it/s]


Images have been successfully sorted into gender and class folders.


In [19]:
import os
import shutil
from tqdm import tqdm

# 원본 이미지 경로와 분류할 경로 설정
source_dir = 'C:/Users/SW/Desktop/데크캠/2024 데이터 크리에이터 캠프 대학부/dataset/validation_image'
destination_dir = 'C:/Users/SW/Desktop/데크캠/2024 데이터 크리에이터 캠프 대학부/dataset/sorted_validation_images'

# 폴더가 없다면 새로 생성
if not os.path.exists(destination_dir):
    os.makedirs(destination_dir)

# 성별 폴더(W, M)를 생성
gender_dirs = {'W': os.path.join(destination_dir, 'W'), 'M': os.path.join(destination_dir, 'M')}
for gender_dir in gender_dirs.values():
    if not os.path.exists(gender_dir):
        os.makedirs(gender_dir)

# os.listdir로 이미지를 리스트로 가져옴
file_list = os.listdir(source_dir)

# tqdm을 사용하여 진행 상태를 표시
for filename in tqdm(file_list, desc="Sorting images"):
    if filename.endswith('.jpg'):
        # 파일명에서 성별과 클래스를 추출
        parts = filename.split('_')
        gender = parts[-1][0]  # 마지막 부분이 'W' 또는 'M'
        class_name = parts[3]  # 네 번째 부분이 클래스명

        # 성별에 따른 폴더 경로
        gender_dir = gender_dirs.get(gender)

        # 클래스별 폴더 경로 설정
        class_dir = os.path.join(gender_dir, class_name)

        # 클래스별 폴더가 없으면 생성
        if not os.path.exists(class_dir):
            os.makedirs(class_dir)

        # 이미지 파일을 클래스별 폴더로 복사 (파일이 없을 때만 복사)
        source_path = os.path.join(source_dir, filename)
        destination_path = os.path.join(class_dir, filename)
        if not os.path.exists(destination_path):  # 파일이 존재하지 않으면 복사
            shutil.copy(source_path, destination_path)

print("Images have been successfully sorted into gender and class folders.")


Sorting images: 100%|███████████████████████████████████████████████████████████████| 951/951 [00:08<00:00, 106.87it/s]


Images have been successfully sorted into gender and class folders.


## resnet-18 기본모델(성별 분류) 1차

1. 학습 및 검증 함수

In [ ]:
# 학습 함수
def train_model(model, dataloaders, criterion, optimizer, num_epochs=25):
    for epoch in range(num_epochs):
        print(f'Epoch {epoch}/{num_epochs - 1}')
        print('-' * 10)

        # 각 epoch은 train과 validation 단계로 나눠짐
        for phase in ['train', 'val']:
            if phase == 'train':
                model.train()  # 모델을 학습 모드로 설정
            else:
                model.eval()  # 모델을 평가 모드로 설정

            running_loss = 0.0
            running_corrects = 0

            # 데이터 로딩
            for inputs, labels in dataloaders[phase]:
                inputs = inputs.to(device)
                labels = labels.to(device)

                # 옵티마이저 초기화
                optimizer.zero_grad()

                # forward
                with torch.set_grad_enabled(phase == 'train'):
                    outputs = model(inputs)
                    _, preds = torch.max(outputs, 1)  # Top-1 예측
                    loss = criterion(outputs, labels)

                    # 학습 단계에서만 backward + optimize 수행
                    if phase == 'train':
                        loss.backward()
                        optimizer.step()

                # 통계 계산
                running_loss += loss.item() * inputs.size(0)
                running_corrects += torch.sum(preds == labels.data)

            epoch_loss = running_loss / len(dataloaders[phase].dataset)
            epoch_acc = running_corrects.double() / len(dataloaders[phase].dataset)

            print(f'{phase} Loss: {epoch_loss:.4f} Acc: {epoch_acc:.4f}')

    return model

# 테스트 함수
def test_model(model, dataloader, device):
    model.eval()  # 모델을 평가 모드로 설정
    running_corrects = 0

    with torch.no_grad():
        for inputs, labels in dataloader:
            inputs = inputs.to(device)
            labels = labels.to(device)

            outputs = model(inputs)
            _, preds = torch.max(outputs, 1)  # Top-1 예측
            running_corrects += torch.sum(preds == labels.data)

    accuracy = running_corrects.double() / len(dataloader.dataset)
    print(f'Test Accuracy (Top-1): {accuracy:.4f}')
    return accuracy

2. 데이터 전처리:
이미지를 224x224로 리사이즈하고, 일반적인 전처리 과정을 적용

3. 학습 및 테스트 실행

In [15]:
# 하이퍼파라미터 설정
num_epochs = 10
learning_rate = 0.001
batch_size = 32

# 디바이스 설정 (GPU 사용 가능하면 GPU로 설정)
device = torch.device("cuda:0" if torch.cuda.is_available() else "cpu")

# 데이터 전처리 (학습 및 검증용)
data_transforms = {
    'train': transforms.Compose([
        transforms.Resize((224, 224)),  # 이미지를 ResNet의 입력 크기로 조정
        transforms.RandomHorizontalFlip(),  # 데이터 증강
        transforms.ToTensor(),
        transforms.Normalize([0.485, 0.456, 0.406], [0.229, 0.224, 0.225])  # ResNet에서 사용된 평균 및 표준편차로 정규화
    ]),
    'val': transforms.Compose([
        transforms.Resize((224, 224)),
        transforms.ToTensor(),
        transforms.Normalize([0.485, 0.456, 0.406], [0.229, 0.224, 0.225])
    ]),
    'test': transforms.Compose([
        transforms.Resize((224, 224)),
        transforms.ToTensor(),
        transforms.Normalize([0.485, 0.456, 0.406], [0.229, 0.224, 0.225])
    ]),
}

# 데이터셋 로드
train_dataset = datasets.ImageFolder('C:/Users/SW/Desktop/데크캠/2024 데이터 크리에이터 캠프 대학부/dataset/sorted_training_images',
                                     transform=data_transforms['train'])
val_dataset = datasets.ImageFolder('C:/Users/SW/Desktop/데크캠/2024 데이터 크리에이터 캠프 대학부/dataset/sorted_validation_images',
                                   transform=data_transforms['val'])

# 데이터로더 준비
dataloaders = {
    'train': DataLoader(train_dataset, batch_size=batch_size, shuffle=True),
    'val': DataLoader(val_dataset, batch_size=batch_size, shuffle=False)
}

# 모델 정의 (간단한 ResNet-18 예시, 전이학습 미사용)
model = models.resnet18(weights=None)  # weights=None으로 전이학습 미사용
num_ftrs = model.fc.in_features
model.fc = nn.Linear(num_ftrs, 2)  # 2개의 성별 클래스 (W, M)를 분류

model = model.to(device)

# 손실 함수 및 옵티마이저 설정
criterion = nn.CrossEntropyLoss()
optimizer = optim.Adam(model.parameters(), lr=learning_rate)

# 학습 실행
trained_model = train_model(model, dataloaders, criterion, optimizer, device, num_epochs=num_epochs)

Epoch 1/10
----------


Train Epoch 1/10: 100%|█| 128/128 [19:53<00:00,  9.32s/batch, accuracy=tensor(0.5317, device='cuda:0', dtype=torch.floa
Val Epoch 1/10: 100%|█| 30/30 [04:32<00:00,  9.08s/batch, accuracy=tensor(0.5605, device='cuda:0', dtype=torch.float64)


val Loss: 0.6777 Acc: 0.5605

Epoch 2/10
----------


Train Epoch 2/10: 100%|█| 128/128 [20:25<00:00,  9.57s/batch, accuracy=tensor(0.5560, device='cuda:0', dtype=torch.floa
Val Epoch 2/10: 100%|█| 30/30 [04:35<00:00,  9.17s/batch, accuracy=tensor(0.6267, device='cuda:0', dtype=torch.float64)


val Loss: 0.6554 Acc: 0.6267

Epoch 3/10
----------


Train Epoch 3/10: 100%|█| 128/128 [21:04<00:00,  9.88s/batch, accuracy=tensor(0.5803, device='cuda:0', dtype=torch.floa
Val Epoch 3/10: 100%|█| 30/30 [04:30<00:00,  9.02s/batch, accuracy=tensor(0.6151, device='cuda:0', dtype=torch.float64)


val Loss: 0.6481 Acc: 0.6151

Epoch 4/10
----------


Train Epoch 4/10: 100%|█| 128/128 [18:48<00:00,  8.82s/batch, accuracy=tensor(0.5749, device='cuda:0', dtype=torch.floa
Val Epoch 4/10: 100%|█| 30/30 [04:17<00:00,  8.58s/batch, accuracy=tensor(0.6299, device='cuda:0', dtype=torch.float64)


val Loss: 0.6474 Acc: 0.6299

Epoch 5/10
----------


Train Epoch 5/10: 100%|█| 128/128 [18:49<00:00,  8.82s/batch, accuracy=tensor(0.6025, device='cuda:0', dtype=torch.floa
Val Epoch 5/10: 100%|█| 30/30 [04:52<00:00,  9.74s/batch, accuracy=tensor(0.5731, device='cuda:0', dtype=torch.float64)


val Loss: 0.6705 Acc: 0.5731

Epoch 6/10
----------


Train Epoch 6/10: 100%|█| 128/128 [21:25<00:00, 10.05s/batch, accuracy=tensor(0.6165, device='cuda:0', dtype=torch.floa
Val Epoch 6/10: 100%|█| 30/30 [04:55<00:00,  9.86s/batch, accuracy=tensor(0.5836, device='cuda:0', dtype=torch.float64)


val Loss: 0.7164 Acc: 0.5836

Epoch 7/10
----------


Train Epoch 7/10: 100%|█| 128/128 [21:42<00:00, 10.18s/batch, accuracy=tensor(0.6103, device='cuda:0', dtype=torch.floa
Val Epoch 7/10: 100%|█| 30/30 [04:34<00:00,  9.16s/batch, accuracy=tensor(0.6309, device='cuda:0', dtype=torch.float64)


val Loss: 0.6368 Acc: 0.6309

Epoch 8/10
----------


Train Epoch 8/10: 100%|█| 128/128 [20:49<00:00,  9.76s/batch, accuracy=tensor(0.6342, device='cuda:0', dtype=torch.floa
Val Epoch 8/10: 100%|█| 30/30 [04:54<00:00,  9.80s/batch, accuracy=tensor(0.6698, device='cuda:0', dtype=torch.float64)


val Loss: 0.6198 Acc: 0.6698

Epoch 9/10
----------


Train Epoch 9/10: 100%|█| 128/128 [21:16<00:00,  9.97s/batch, accuracy=tensor(0.6521, device='cuda:0', dtype=torch.floa
Val Epoch 9/10: 100%|█| 30/30 [05:02<00:00, 10.07s/batch, accuracy=tensor(0.6488, device='cuda:0', dtype=torch.float64)


val Loss: 0.6375 Acc: 0.6488

Epoch 10/10
----------


Train Epoch 10/10: 100%|█| 128/128 [20:29<00:00,  9.60s/batch, accuracy=tensor(0.6654, device='cuda:0', dtype=torch.flo
Val Epoch 10/10: 100%|█| 30/30 [04:37<00:00,  9.26s/batch, accuracy=tensor(0.5047, device='cuda:0', dtype=torch.float64

val Loss: 0.8223 Acc: 0.5047



In [16]:
# 테스트용 데이터 로드 (테스트 시에만 사용)
test_dataset = datasets.ImageFolder('C:/Users/SW/Desktop/데크캠/2024 데이터 크리에이터 캠프 대학부/dataset/sorted_validation_images',
                                    transform=data_transforms['test'])
test_loader = DataLoader(test_dataset, batch_size=batch_size, shuffle=False)

# 테스트 실행
test_model(trained_model, test_loader, device)

Testing: 100%|███████████████████████████████████████████████████████| 30/30 [04:40<00:00,  9.36s/batch, accuracy=50.5]

Accuracy on the test set: 50.47%


-정리

num_epochs = 10

learning_rate = 0.001

batch_size = 32

총 4시간 소요, accuracy=50.5

문제점: validation/test 파일 분리를 하지 못해서 test결과가 정확하지 못함,

원본 이미지파일이 얼굴을 전부 가려서 성별 구별이 쉽지않음 + 배경이 너무 다채로움

## 2차: object detection, image cropping 전처리 실시